Pr branch

TasnimAwashreh opened the pull request.

TasnimAwashreh committed:
- Implemented Observer Pattern for the subject and bots

TasnimAwashreh committed:
- Added Configuration
- Moved config file to the project root
- Fixed the structure of the project
- Added a constants file for strings
- added an input parser where the user can enter multiple lines
- Added a format reader that chooses between xml or json depending on the user's input

TasnimAwashreh committed:
- New branch for PR

TasnimAwashreh committed:
- Renamed Subject to WeatherStation
- changed config.json to appsettings.json
- Moved WeatherData to Domain
- Bots now check if theyre enabled to update

TasnimAwashreh committed:
- Added DTOs

TasnimAwashreh committed:
- Surrounded readers with try catch

TasnimAwashreh committed:
- Updated Reader classes

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Readers/JsonFormatReader.cs:
```diff
@@ -0,0 +1,20 @@
+            {
+                return JsonSerializer.Deserialize<WeatherData>(userInput);
+            }
+            catch (Exception ex)
```
we need to specify the exception that related to invalid format, and then we shouldn't just swallow the exception and return null, we should throw an appropriate exception indicating invalid exception

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Readers/XMLFormatReader.cs:
```diff
@@ -0,0 +1,23 @@
+                WeatherData data = (WeatherData)serializer.Deserialize(reader);
+                return data;
+            }
+            catch (Exception ex)
```
same comment as above

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/DTOs/BotDTO.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace WeatherMonitoringAndReportingService.Logic.DTOs
+{
+    public class BotDTO
```
a better name is `BotConfig`. And `Configuration` for the folder

RandaZraik replied:
same comment for the other config models

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Services/ReaderService.cs:
```diff
@@ -0,0 +1,29 @@
+            _reader = reader;
+        }
+
+        public void ChooseReader(string input)
```
we should use the [strategy pattern](https://refactoring.guru/design-patterns/strategy/csharp/example)

RandaZraik commented:
Original [PR #1](https://github.com/TasnimAwashreh/Patterns_WeatherMonitoringAndReportingService/pull/1)

TasnimAwashreh committed:
- Removed the try-catch to allow exception to bubble up and print out in the App layer
- renamed DTOs to configs + renamed folder to Configuration
- Updated strategy pattern

TasnimAwashreh committed:
- Added set reader in service
